[Form Lib] getFields returns typed FieldHooks

rylnd opened the pull request:
🚧 🚧 

We were previously losing our type information when using the
`getFields` method; it now returns a properly typed map of `FieldHook`s asvalues.

### Checklist

Check the PR satisfies following conditions. 

Reviewers should verify this PR satisfies this list as well.

- [ ] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/src/platform/packages/shared/kbn-i18n/README.md)
- [ ] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [ ] If a plugin configuration key changed, check if it needs to be allowlisted in the cloud and added to the [docker list](https://github.com/elastic/kibana/blob/main/src/dev/build/tasks/os_packages/docker_generator/resources/base/bin/kibana-docker)
- [ ] This was checked for breaking HTTP API changes, and any breaking changes have been approved by the breaking-change committee. The `release_note:breaking` label should be applied in these situations.
- [ ] [Flaky Test Runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) was used on any tests changed
- [ ] The PR  description includes the appropriate Release Notes section, and the correct `release_note:*` label is applied per the [guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)
- [ ] Review the [backport guidelines](https://docs.google.com/document/d/1VyN5k91e5OVumlc0Gb9RPa3h1ewuPE705nRtioPiTvY/edit?usp=sharing) and apply applicable `backport:*` labels.

### Identify risks

Does this PR introduce any risks? For example, consider risks like hard to test bugs, performance regression, potential of data loss.

Describe the risk, its severity, and mitigation for each identified risk. Invite stakeholders and evaluate how to proceed before merging.

- [ ] [See some risk examples](https://github.com/elastic/kibana/blob/main/RISK_MATRIX.mdx)
- [ ] ...

rylnd committed:
Leverage types in  FormHook's `getFields` method

We were previously losing our type information when using the
`getFields` method; it now returns properly typed `FieldHook`s as
values.

rylnd committed:
WIP: cleaning up now-unnecessary type assertions

rylnd committed:
Fix incorrect removal of type information

Previously, line 57 contained:

* A destructuring
* A type assertion
* A property access on the RHS _after_ the type assertion

I had missed the third part, which caused me to add those TODO comments.
This should now be fixed.

elasticmachine commented:
## :broken_heart: Build Failed
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/367920)
* Commit: 9a8bb85d1d21eb73d918f72b086c4c442dc69141

### Failed CI Steps
* [:pipeline::magic_wand: Run Dynamic Pipeline: .buildkite/scripts/pipelines/pull_request/pipeline.sh](https://buildkite.com/elastic/kibana-pull-request/builds/367920#019ae077-e228-424a-92e7-638d82597b8b)

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/367920","number":367920,"commit":"9a8bb85d1d21eb73d918f72b086c4c442dc69141"}],"number":367920}
buildkite-pr-comment-->

rylnd committed:
Use Record as internal map in useForm

Since we assign arbitrary keys to the `fieldsRefs` ref, typescript is no
longer confident that every key used is `keyof T`. By using a Record
internally and then casting at API boundaries, we can sidestep this
concern.

rylnd committed:
Prevent calling method on optional FieldHook

The form data states that this field is possibly undefined, therefore
with proper types, unconditionally calling `.reset` on the corresponding
FieldHook is a potential runtime error.

rylnd committed:
Remove more unnecessary type assertions

rylnd committed:
Assert presence of name field in submit handler

While the form data is typed with `name` as an optional field, in this
function we are explicitly setting the `name` to have the same value as
the `title`.

elasticmachine commented:
## :hourglass_flowing_sand: Build in-progress, with failures
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/368642)
* Commit: b7eaee815bbf07fe84d21462caa580131a245a63

### Failed CI Steps
* [FTR Configs #135](https://buildkite.com/elastic/kibana-pull-request/builds/368642#019ae585-5f34-4a97-9550-9bf351bc28da)
* [FTR Configs #135](https://buildkite.com/elastic/kibana-pull-request/builds/368642#019ae5ab-cd2e-4f55-9895-f52312cb5dce)

### Test Failures
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/368642#019ae585-5f34-4a97-9550-9bf351bc28da) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/368642/jobs/019ae585-5f34-4a97-9550-9bf351bc28da/artifacts/019ae5ab-9209-4905-ac04-870e1be49ce8) FTR Configs #<span></span>135 / Maps endpoints apis search ES|QL should return getValues response in expected shape
* [[job]](https://buildkite.com/elastic/kibana-pull-request/builds/368642#019ae5ab-cd2e-4f55-9895-f52312cb5dce) [[logs]](https://buildkite.com/organizations/elastic/pipelines/kibana-pull-request/builds/368642/jobs/019ae5ab-cd2e-4f55-9895-f52312cb5dce/artifacts/019ae5b1-e137-4454-9996-5cc7b74df474) FTR Configs #<span></span>135 / Maps endpoints apis search ES|QL should return getValues response in expected shape

### History
* :broken_heart: [Build #368068](https://buildkite.com/elastic/kibana-pull-request/builds/368068) failed b885a71954e06f56b58bab782a465fc368d62371
* :broken_heart: [Build #368004](https://buildkite.com/elastic/kibana-pull-request/builds/368004) failed 0824856ddddee8bdad3ebbe672bbaa50cc86e624
* :broken_heart: [Build #367920](https://buildkite.com/elastic/kibana-pull-request/builds/367920) failed 9a8bb85d1d21eb73d918f72b086c4c442dc69141

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368642","number":368642,"commit":"b7eaee815bbf07fe84d21462caa580131a245a63"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368068","number":368068,"commit":"b885a71954e06f56b58bab782a465fc368d62371"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368004","number":368004,"commit":"0824856ddddee8bdad3ebbe672bbaa50cc86e624"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/367920","number":367920,"commit":"9a8bb85d1d21eb73d918f72b086c4c442dc69141"}],"number":368642}
buildkite-pr-comment-->

rylnd committed:
Merge branch 'main' into typed-fieldsmap

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/369175)
* Commit: 4566d61a50ed8f9244021c1ad60269b6c52ee593

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/f93bb7ca0b97a925793954fca6e8ea67a6ed20e1) | [after](https://github.com/elastic/kibana/commit/4566d61a50ed8f9244021c1ad60269b6c52ee593) | diff |
| --- | --- | --- | --- |
| `securitySolution` | 11.1MB | 11.1MB | +26.0B |



### History
* :broken_heart: [Build #368769](https://buildkite.com/elastic/kibana-pull-request/builds/368769) failed b7eaee815bbf07fe84d21462caa580131a245a63
* :broken_heart: [Build #368642](https://buildkite.com/elastic/kibana-pull-request/builds/368642) failed b7eaee815bbf07fe84d21462caa580131a245a63
* :broken_heart: [Build #368068](https://buildkite.com/elastic/kibana-pull-request/builds/368068) failed b885a71954e06f56b58bab782a465fc368d62371
* :broken_heart: [Build #368004](https://buildkite.com/elastic/kibana-pull-request/builds/368004) failed 0824856ddddee8bdad3ebbe672bbaa50cc86e624
* :broken_heart: [Build #367920](https://buildkite.com/elastic/kibana-pull-request/builds/367920) failed 9a8bb85d1d21eb73d918f72b086c4c442dc69141

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/369175","number":369175,"commit":"4566d61a50ed8f9244021c1ad60269b6c52ee593"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368769","number":368769,"commit":"b7eaee815bbf07fe84d21462caa580131a245a63"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368642","number":368642,"commit":"b7eaee815bbf07fe84d21462caa580131a245a63"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368068","number":368068,"commit":"b885a71954e06f56b58bab782a465fc368d62371"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/368004","number":368004,"commit":"0824856ddddee8bdad3ebbe672bbaa50cc86e624"},{"buildStatus":{"state":"failed","success":false,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/367920","number":367920,"commit":"9a8bb85d1d21eb73d918f72b086c4c442dc69141"}],"number":369175}
buildkite-pr-comment-->

rylnd committed:
Merge branch 'main' into typed-fieldsmap
